src/datasets/arrow_dataset.py

```diff
@@ -159,6 +159,131 @@ def version(self):
+            else:
+                dtype_str = col_feature.dtype
+            if dtype_str.startswith("int") or dtype_str.startswith("uint"):
+                dtype = tf.int32
+            elif dtype_str.startswith("float"):
+                dtype = tf.float32
+            else:
+                raise ValueError(f"Could not convert datatype {dtype_str} in column {column}!")
+
+            if isinstance(col_feature, Value):
+                shape = [batch_size]
+            elif isinstance(col_feature, _ArrayXD):
+                shape = [batch_size] + list(col_feature.shape)
+            elif isinstance(col_feature, Sequence):
+                shape = [batch_size, col_feature.length]
+            else:
+                raise ValueError(
+                    f"Couldn't parse feature {column} with type {type(col_feature)}! "
+                    "This may indicate a column was included with an unusual datatype "
+                    "that we were unable to process correctly. "
+                    "If you're getting this error with one of our datasets, and you're "
+                    "sure the column should be convertable to tf.Tensor, please "
+                    "file an issue at github.com/huggingface/datasets and tag "
+                    "@rocketknight1!"
+                )
+            shape = [dim if dim != -1 else None for dim in shape]
+
+            signatures[column] = tf.TensorSpec(shape=shape, dtype=dtype)
+        return signatures
+
+    def to_tf_dataset(self, columns, batch_size, shuffle, collate_fn=None, label_cols=None, pad_to=0, prefetch=True):
+        import tensorflow as tf
+
+        if label_cols is None:
+            label_cols = []
+        elif isinstance(label_cols, str):
+            label_cols = [label_cols]
+        elif len(set(label_cols)) < len(label_cols):
+            raise ValueError("List of label_cols contains duplicates!")
+        if not columns:
+            raise ValueError("Need to specify at least one column!")
+        elif isinstance(columns, str):
+            columns = [columns]
+        elif len(set(columns)) < len(columns):
+            raise ValueError("List of columns contains duplicates!")
+        if pad_to > 0 and collate_fn is not None:
+            raise ValueError(
+                "pad_to cannot be used with a custom collate_fn - you should modify your collate_fn instead!"
+            )
+        if label_cols is not None:
+            cols_to_retain = list(set(columns + label_cols))
+        else:
+            cols_to_retain = columns
+        for col in cols_to_retain:
+            if col not in self.features:
+                raise ValueError(f"Couldn't find column {col} in dataset!")
+        dataset = self.remove_columns([col for col in self.features if col not in cols_to_retain])
+        gen_signature = self._get_output_signature(dataset, batch_size)
+        num_batches = len(dataset) // batch_size  # Because we drop the remainder
```

Comment:
You should had a flag to enable/disable this: you never want to drop the remainder for evaluation, only training.

Reply:
Will do!

Reply:
Done!